cmd/bisync/bisync_test.go

```diff
@@ -0,0 +1,1249 @@
+func (b *bisyncTest) mangleResult(dir, file string, golden bool) string {
+	buf, err := ioutil.ReadFile(filepath.Join(dir, file))
+	require.NoError(b.t, err)
+	text := string(buf)
+
+	switch fileType(strings.TrimSuffix(file, ".sav")) {
+	case "queue":
+		lines := strings.Split(text, eol)
+		sort.Strings(lines)
+		return joinLines(lines)
+	case "listing":
+		return mangleListing(text, golden)
+	case "log":
+		// fall thru
+	default:
+		return text
+	}
+
+	// Adapt log lines to the golden way.
+	lines := strings.Split(string(buf), eol)
+	pathReplacer := b.newReplacer(true)
+
+	rep := logReplacements
+	if b.testCase == "dry_run" {
+		rep = append(rep, dryrunReplacements...)
+	}
+	repFrom := make([]*regexp.Regexp, len(rep)/2)
+	repTo := make([]string, len(rep)/2)
+	for i := 0; i < len(rep); i += 2 {
+		repFrom[i/2] = regexp.MustCompile(rep[i])
+		repTo[i/2] = rep[i+1]
+	}
+
+	hoppers := make([]*regexp.Regexp, len(logHoppers))
+	dampers := make([][]string, len(logHoppers))
+	for i, regex := range logHoppers {
+		hoppers[i] = regexp.MustCompile("^" + regex + "$")
+	}
+
+	// The %q format doubles backslashes, hence "{1,2}"
+	regexBackslash := regexp.MustCompile(`\\{1,2}`)
+
+	emptyCount := 0
+	maxEmpty := 0
+	if b.golden {
+		maxEmpty = 2
+	}
+
+	result := make([]string, 0, len(lines))
+	for _, s := range lines {
+		// Adapt file paths
+		s = pathReplacer.Replace(strings.TrimSpace(s))
+
+		// Apply regular expression replacements
+		for i := 0; i < len(repFrom); i++ {
+			s = repFrom[i].ReplaceAllString(s, repTo[i])
+		}
+		s = strings.TrimSpace(s)
+		if s == dropMe {
+			continue
+		}
+
+		if fixSlash && regexFixSlash.MatchString(s) {
+			s = regexBackslash.ReplaceAllString(s, "/")
+		}
+
+		// Sort consecutive groups of naturally unordered lines.
+		// Any such group must end before the log ends or it might be lost.
+		absorbed := false
+		for i := 0; i < len(dampers); i++ {
+			match := false
+			if s != "" && !absorbed {
+				match = hoppers[i].MatchString(s)
+			}
+			if match {
+				dampers[i] = append(dampers[i], s)
+				absorbed = true
+			} else if len(dampers[i]) > 0 {
+				sort.Strings(dampers[i])
+				result = append(result, dampers[i]...)
+				dampers[i] = nil
+			}
+		}
+		if absorbed {
+			continue
+		}
+
+		// Skip empty lines unless storing to golden
+		if s == "" {
+			if emptyCount < maxEmpty {
+				result = append(result, "")
+			}
+			emptyCount++
+			continue
+		}
+		result = append(result, s)
+		emptyCount = 0
+	}
+
+	return joinLines(result)
+}
+
+// mangleListing sorts listing lines before comparing.
+func mangleListing(text string, golden bool) string {
+	lines := strings.Split(text, eol)
+
+	hasHeader := len(lines) > 0 && strings.HasPrefix(lines[0], bisync.ListingHeader)
+	if hasHeader {
+		lines = lines[1:]
+	}
+
+	// Split line in 4 groups:    (flag, size)(hash.)( .id., .......modtime....... )(name).
+	regex := regexp.MustCompile(`^([^ ] +\d+ )([^ ]+)( [^ ]+ [\d-]+T[\d:.]+[\d+-]+ )(".+")$`)
```

Comment:
@ncw 

`mangleListing` is a very far relative of [fstest.CompareItems](https://github.com/rclone/rclone/blob/v1.55.1/fstest/fstest.go#L356). It already has ability to ignore hashsums on demand. It still needs ability to compare times with per-backend tolerance or even ignore times if unsupported. Having that would let us run tests against backend without time support and probably lift the corresponding ban.

Reply:
Tentatively this can be implemented as:
1. parse out the time filed and convert to `time.Time`
2. round time down to _doubled_ tolerance
3. put time field back into string
4. feed the result to `diff`.
I hope it does not sound scary, but it would be part of the proposed approach to testing.

Reply:
Why not just parse and compare with tolerance?
Because `mangleResult` is two-fold. Besides comparing it can prepare a test state snapshot for storing in golden directory, which later can provide a bigger picture of what's going on.